pkg/sql/opt/xform/select_funcs.go

```diff
@@ -471,35 +484,67 @@ func (c *CustomFuncs) GenerateConstrainedScans(
 					continue
 				}
 				for j := 0; j < optionalConstraints.Length(); j++ {
-					if combinedConstraint.Contains(c.e.ctx, c.e.evalCtx, optionalConstraints.Constraint(j)) {
+					if cand.constraint.Contains(c.e.ctx, c.e.evalCtx, optionalConstraints.Constraint(j)) {
 						return
 					}
 				}
 			}
 		}
+		candidates = append(candidates, cand)
+	})
+
+	// Prune candidates with constraints that contain other constraints, unless
+	// one avoids an index join. These constraints are strictly worse than the
+	// constraints they contain. For example, if we have the two constraints:
+	//
+	// /1/2: [/0 - /0]
+	// /1/2: [/0/100 - /0/101]
+	//
+	// The second constraint is more selective than the first. There is no
+	// reason to ever scan the first unless it is covering and avoids an index
+	// join.
+	var pruned intsets.Fast
+	for i := 0; i < len(candidates); i++ {
+		for j := i + 1; j < len(candidates); j++ {
+			cand1, cand2 := candidates[i], candidates[j]
+			if cand1.isCovering != cand2.isCovering {
```

Comment:
We can probably do something smarter here. E.g., if cand1 is not covering and cand2 is covering, we still want to prune cand1 if it's contraint contains cand2's constraint.